jax/_src/numpy/lax_numpy.py

```diff
@@ -3417,17 +3417,22 @@ def _take(a, indices, axis: Optional[int] = None, out=None, mode=None):
 
 def _normalize_index(index, axis_size):
   """Normalizes an index value in the range [-N, N) to the range [0, N)."""
-  if issubdtype(_dtype(index), np.unsignedinteger):
-    return index
-  if core.is_constant_dim(axis_size):
-    axis_size_val = _lax_const(index, axis_size)
-  else:
-    axis_size_val = lax.convert_element_type(core.dimension_as_value(axis_size),
-                                             _dtype(index))
-  return lax.select(
-    lax.lt(index, _lax_const(index, 0)),
-    lax.add(index, axis_size_val),
-    index)
+  try:
+    idx = core.concrete_or_error(operator.index, index, "normalize index")
+    ndim = core.concrete_or_error(operator.index, axis_size, "normalize axis_size")
```

Comment:
I think you can simplify this: concrete tracers implement the `__index__` method, so if the concretization error is not going to be user-facing, there's no benefit to calling `core.concrete_or_error`. Instead you could just call `operator.index` directly, and you'd only have to catch the `TypeError`.

Reply:
Actually, I just realized that if `operator.index` is called on a non-concrete tracer, it will result in a `TracerIntegerConversionError`, so you'll have to also catch that.

Reply:
Great, this is much nicer! `TracerIntegerConversionError` is a subclass of `TypeError`, so this will already be caught together with errors from `operator.index`.
I changed it in the latest commit.

Reply:
Hi! - @jakevdp , concrete tracer will raise exception on `__index__` but not `int`!
```python
from jax import grad
grad(lambda x: print(x.aval) or print(int(x)) or print(x.__index__()) or x)(1.0)
# ConcreteArray(1.0, dtype=float32, weak_type=True)
# 1
# jax._src.errors.TracerIntegerConversionError: The __index__() method was called on the JAX Tracer object 
# Traced<ConcreteArray(1.0, dtype=float32, weak_type=True)>with<JVPTrace(level=2/0)
```